packages/grpc-js/src/xds-client.ts

```diff
@@ -0,0 +1,389 @@
+      node: this.node!,
+      type_url: EDS_TYPE_URL,
+      resource_names: Array.from(this.endpointWatchers.keys()),
+      response_nonce: this.lastEdsNonce,
+      version_info: this.lastEdsVersionInfo,
+      error_detail: {
+        message,
+      },
+    });
+  }
+
+  /**
+   * Validate the ClusterLoadAssignment object by these rules:
+   * https://github.com/grpc/proposal/blob/master/A27-xds-global-load-balancing.md#clusterloadassignment-proto
+   * @param message
+   */
+  private validateEdsResponse(message: ClusterLoadAssignment__Output): boolean {
+    for (const endpoint of message.endpoints) {
+      for (const lb of endpoint.lb_endpoints) {
+        const socketAddress = lb.endpoint?.address?.socket_address;
+        if (!socketAddress) {
+          return false;
+        }
+        if (socketAddress.port_specifier !== 'port_value') {
+          return false;
+        }
+        if (!(isIPv4(socketAddress.address) || isIPv6(socketAddress.address))) {
+          return false;
+        }
+      }
+    }
+    return true;
+  }
+
+  private handleEdsResponse(message: ClusterLoadAssignment__Output) {
+    const watchers = this.endpointWatchers.get(message.cluster_name) ?? [];
+    for (const watcher of watchers) {
+      watcher.onValidUpdate(message);
+    }
+  }
+
+  private updateEdsNames() {
+    if (this.adsCall) {
+      this.adsCall.write({
+        node: this.node!,
+        type_url: EDS_TYPE_URL,
+        resource_names: Array.from(this.endpointWatchers.keys()),
+        response_nonce: this.lastEdsNonce,
+        version_info: this.lastEdsVersionInfo,
+      });
+    }
+  }
+
+  private reportStreamError(status: StatusObject) {
+    for (const watcherList of this.endpointWatchers.values()) {
+      for (const watcher of watcherList) {
+        watcher.onTransientError(status);
+      }
+    }
+    // Also do the same for other types of watchers when those are implemented
+  }
+
+  addEndpointWatcher(
+    edsServiceName: string,
+    watcher: Watcher<ClusterLoadAssignment__Output>
+  ) {
+    trace('Watcher added for endpoint ' + edsServiceName);
+    let watchersEntry = this.endpointWatchers.get(edsServiceName);
+    let addedServiceName = false;
+    if (watchersEntry === undefined) {
+      addedServiceName = true;
+      watchersEntry = [];
+      this.endpointWatchers.set(edsServiceName, watchersEntry);
+    }
+    watchersEntry.push(watcher);
+    if (addedServiceName) {
+      this.updateEdsNames();
+    }
+  }
+
+  removeEndpointWatcher(
+    edsServiceName: string,
+    watcher: Watcher<ClusterLoadAssignment__Output>
+  ) {
+    trace('Watcher removed for endpoint ' + edsServiceName);
+    const watchersEntry = this.endpointWatchers.get(edsServiceName);
+    let removedServiceName = false;
+    if (watchersEntry !== undefined) {
+      const entryIndex = watchersEntry.indexOf(watcher);
+      if (entryIndex >= 0) {
+        watchersEntry.splice(entryIndex, 1);
+      }
+      if (watchersEntry.length === 0) {
```

Comment:
Does comparing with 0 have to be `===`? I guess it doesn't hurt.

Reply:
The convention is that all comparisons should be done using `===`.